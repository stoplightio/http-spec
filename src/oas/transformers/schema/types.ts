import type { JSONSchema7 } from 'json-schema';
import type { SchemaObject } from 'openapi3-ts/src/model/OpenApi';
import type { Schema } from 'swagger-schema-official';

export type OASSchemaObject = SchemaObject | Schema;

export type Converter = (schema: OASSchemaObject | JSONSchema7) => void;
