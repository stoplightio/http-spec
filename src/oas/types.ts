import { DeepPartial, IHttpOperation, IHttpService } from '@stoplight/types';
import { OpenAPIObject } from 'openapi3-ts';
import { Spec } from 'swagger-schema-official';

export const enum OasVersion {
  OAS2 = 2,
  OAS3 = 3,
}

/**
 * Service
 */

export interface ITransformServiceOpts<T> {
  document: DeepPartial<T>;
}

export type HttpServiceTransformer<T> = (opts: T) => IHttpService;

export type Oas2TransformServiceOpts = ITransformServiceOpts<Spec>;
export type Oas3TransformServiceOpts = ITransformServiceOpts<OpenAPIObject>;
export type Oas2HttpServiceTransformer = HttpServiceTransformer<Oas2TransformServiceOpts>;
export type Oas3HttpServiceTransformer = HttpServiceTransformer<Oas3TransformServiceOpts>;

/**
 * Operation
 */

export interface ITransformOperationOpts<T> {
  document: DeepPartial<T>;
  path: string;
  method: string;
}

export type HttpOperationTransformer<T> = (opts: T) => IHttpOperation;

export type Oas2TransformOperationOpts = ITransformOperationOpts<Spec>;
export type Oas3TransformOperationOpts = ITransformOperationOpts<OpenAPIObject>;
export type Oas2HttpOperationTransformer = HttpOperationTransformer<Oas2TransformOperationOpts>;
export type Oas3HttpOperationTransformer = HttpOperationTransformer<Oas3TransformOperationOpts>;
