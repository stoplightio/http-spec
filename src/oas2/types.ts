import type { DeepPartial } from '@stoplight/types';
import type { Spec } from 'swagger-schema-official';

import { TranslateFunction } from '../types';

export type Oas2TranslateFunction<P extends unknown[], R extends unknown = unknown> = TranslateFunction<
  DeepPartial<Spec>,
  P,
  R
>;
