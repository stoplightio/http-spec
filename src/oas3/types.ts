import { DeepPartial } from '@stoplight/types';
import { OpenAPIObject as _OpenAPIObject } from 'openapi3-ts';
import { PathItemObject } from 'openapi3-ts/src/model/OpenApi';
import { ISpecificationExtension } from 'openapi3-ts/src/model/SpecificationExtension';

import { TranslateFunction } from '../types';

export type Oas3TranslateFunction<P extends unknown[], R extends unknown = unknown> = TranslateFunction<
  DeepPartial<OpenAPIObject>,
  P,
  R
>;

export interface OpenAPIObject extends _OpenAPIObject {
  webhooks?: WebhooksObject;
}

export interface WebhooksObject extends ISpecificationExtension {
  [name: string]: PathItemObject;
}
