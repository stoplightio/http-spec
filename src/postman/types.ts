import { Collection } from 'postman-collection';
import { HttpOperationTransformer, ITransformOperationOpts } from '../types';

export type PostmanCollectionHttpOperationTransformer = HttpOperationTransformer<{
  document: Collection;
  path: string;
  method: string;
}>;
