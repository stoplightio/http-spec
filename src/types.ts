import { DeepPartial, IHttpOperation, IHttpService } from '@stoplight/types';

export interface ITransformServiceOpts<T> {
  document: DeepPartial<T>;
}

export type HttpServiceTransformer<T> = (opts: T) => IHttpService;

export interface ITransformOperationOpts<T> {
  document: DeepPartial<T>;
  path: string;
  method: string;
}

export type HttpOperationTransformer<T> = (opts: T) => IHttpOperation;
