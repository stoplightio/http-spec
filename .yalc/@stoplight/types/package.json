{
  "name": "@stoplight/types",
  "version": "0.0.0+61d12013",
  "description": "Common typings for the Stoplight ecosystem.",
  "keywords": [],
  "sideEffects": false,
  "homepage": "https://github.com/stoplightio/types",
  "bugs": "https://github.com/stoplightio/types/issues",
  "author": "Stoplight <support@stoplight.io>",
  "repository": {
    "type": "git",
    "url": "https://github.com/stoplightio/types.git"
  },
  "main": "dist/index.js",
  "license": "Apache-2.0",
  "files": [
    "dist/*"
  ],
  "engines": {
    "node": ">=8"
  },
  "scripts": {
    "build": "tsc -p ./tsconfig.json -d",
    "commit": "git-cz",
    "lint": "tslint -c tslint.json 'src/**/*.ts?'",
    "lint.fix": "yarn lint --fix",
    "release": "semantic-release",
    "test": "jest",
    "test.prod": "yarn lint && yarn test --coverage --no-cache",
    "test.update": "yarn test --updateSnapshot",
    "test.watch": "yarn test --watch"
  },
  "dependencies": {
    "@types/json-schema": "^7.0.4",
    "utility-types": "^3.10.0"
  },
  "lint-staged": {
    "*.ts": [
      "tslint -c tslint.json --fix",
      "git add"
    ]
  },
  "husky": {
    "hooks": {
      "commit-msg": "commitlint -E HUSKY_GIT_PARAMS",
      "pre-commit": "lint-staged"
    }
  },
  "config": {
    "commitizen": {
      "path": "node_modules/cz-conventional-changelog"
    }
  },
  "commitlint": {
    "extends": [
      "@commitlint/config-conventional"
    ]
  },
  "yalcSig": "61d120131f954de9c167a4a88ea3c8bc"
}
