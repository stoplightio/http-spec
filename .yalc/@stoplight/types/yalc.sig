7dd53c0f6cd0fa550fc0ed26f5e42e94