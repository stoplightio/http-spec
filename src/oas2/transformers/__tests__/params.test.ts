import { FormDataParameter, QueryParameter } from 'swagger-schema-official';

import { HttpParamStyles } from '@stoplight/types';
import {
  translateFromFormDataParameters,
  translateToBodyParameter,
  translateToHeaderParam,
  translateToHeaderParams,
  translateToPathParameter,
  translateToQueryParameter,
} from '../params';

describe('params.translator', () => {
  let consumes = ['*'];

  describe('translateToHeaderParam', () => {
    test('should translate header param', () => {
      expect(
        translateToHeaderParam({
          required: true,
          name: 'name',
          type: 'integer',
          in: 'header',
          minimum: 12,
          description: 'desc',
        }),
      ).toMatchSnapshot();
    });
  });

  describe('translateToHeaderParams', () => {
    test('should translate empty dictionary to empty array', () => {
      expect(translateToHeaderParams({})).toMatchSnapshot();
    });

    test('should translate to simple header param', () => {
      expect(
        translateToHeaderParams({
          'header-name': {
            description: 'a description',
            type: 'string',
          },
        }),
      ).toMatchSnapshot();
    });

    test('should translate to multiple header params', () => {
      expect(
        translateToHeaderParams({
          'header-name': {
            description: 'a description',
            type: 'string',
          },
          'plain-tex': {
            description: 'another description',
            type: 'string',
          },
        }),
      ).toMatchSnapshot();
    });
  });

  describe('translateToBodyParameter', () => {
    test('should translate to body parameter', () => {
      expect(
        translateToBodyParameter(
          {
            in: 'body',
            name: 'name',
            required: true,
            description: 'descr',
            schema: {
              format: 'e-mail',
            },
          },
          consumes,
        ),
      ).toMatchSnapshot();
    });

    test('should preserve readOnly flag in schema', () => {
      expect(
        translateToBodyParameter(
          {
            in: 'body',
            name: 'name',
            required: true,
            description: 'descr',
            schema: {
              readOnly: true,
            },
          },
          consumes,
        ),
      ).toEqual(
        expect.objectContaining({
          contents: expect.arrayContaining([expect.objectContaining({ schema: { readOnly: true } })]),
        }),
      );
    });

    test('given x-examples should translate to body parameter with multiple examples', () => {
      expect(
        translateToBodyParameter(
          {
            in: 'body',
            name: 'name',
            required: true,
            description: 'descr',
            schema: {
              format: 'e-mail',
            },
            // @ts-ignore: "x-examples" isn't technically supported by the OAS2 spec
            'x-examples': {
              'example-1': {
                hello: 'world',
              },
              'example-2': {
                foo: 'bar',
              },
            },
          },
          consumes,
        ),
      ).toMatchSnapshot();
    });
  });

  describe('translateToFormDataParameter', () => {
    beforeAll(() => {
      consumes = ['application/x-www-form-urlencoded', 'multipart/form-data'];
    });

    const formDataParameterString: FormDataParameter = {
      in: 'formData',
      name: 'str',
      type: 'string',
      format: 'date',
      description: 'desc',
      required: true,
      allowEmptyValue: false,
      default: '25-07-2019',
    };

    const formDataParameterArray: FormDataParameter = {
      in: 'formData',
      name: 'arr',
      type: 'array',
      description: 'desc',
      collectionFormat: 'multi',
      items: {
        type: 'number',
      },
      required: true,
      maxItems: 10,
      minItems: 1,
    };

    const formDataParameterInteger: FormDataParameter = {
      in: 'formData',
      name: 'int',
      type: 'integer',
      description: 'desc',
      required: true,
      maximum: 3,
      minimum: 0,
    };

    test('converts parameters into schema', () => {
      const expectedContent = {
        encodings: [
          {
            explode: true,
            property: 'arr',
            style: 'form',
          },
        ],
        schema: {
          properties: {
            arr: {
              description: 'desc',
              items: {
                type: 'number',
              },
              maxItems: 10,
              minItems: 1,
              type: 'array',
            },
            int: {
              description: 'desc',
              maximum: 3,
              minimum: 0,
              type: 'integer',
            },
            str: {
              allowEmptyValue: false,
              minLength: 1,
              default: '25-07-2019',
              description: 'desc',
              format: 'date',
              type: 'string',
            },
          },
          required: ['str', 'arr', 'int'],
          type: 'object',
        },
      };

      expect(
        translateFromFormDataParameters(
          [formDataParameterString, formDataParameterArray, formDataParameterInteger],
          consumes,
        ),
      ).toEqual({
        contents: [
          Object.assign({}, expectedContent, { mediaType: 'application/x-www-form-urlencoded' }),
          Object.assign({}, expectedContent, { mediaType: 'multipart/form-data' }),
        ],
      });
    });
  });

  describe('translateToQueryParameter', () => {
    const parameter: QueryParameter = {
      required: true,
      description: 'descr',
      name: 'name',
      in: 'query',
      type: 'number',
      allowEmptyValue: true,
    };

    test.each([
      { oasStyle: 'pipes', expectedStyle: HttpParamStyles.PipeDelimited },
      { oasStyle: 'ssv', expectedStyle: HttpParamStyles.SpaceDelimited },
      { oasStyle: 'csv', expectedStyle: HttpParamStyles.CommaDelimited },
      { oasStyle: 'multi', expectedStyle: HttpParamStyles.Form },
      { oasStyle: 'nasino', expectedStyle: HttpParamStyles.Form },
    ])('translate style: %s', ({ oasStyle, expectedStyle }) => {
      expect(translateToQueryParameter({ ...parameter, collectionFormat: oasStyle } as QueryParameter)).toHaveProperty(
        'style',
        expectedStyle,
      );
    });
  });

  describe('translateToPathParameter', () => {
    test('should translate', () => {
      expect(
        translateToPathParameter({
          required: true,
          description: 'descr',
          name: 'name',
          in: 'path',
          type: 'string',
        }),
      ).toMatchSnapshot();
    });
  });
});
