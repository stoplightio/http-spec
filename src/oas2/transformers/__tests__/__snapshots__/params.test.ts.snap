// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`params.translator translateToBodyParameter given x-examples should translate to body parameter with multiple examples 1`] = `
Object {
  "contents": Array [
    Object {
      "examples": Array [
        Object {
          "id": Any<String>,
          "key": "example-1",
          "value": Object {
            "hello": "world",
          },
        },
        Object {
          "id": Any<String>,
          "key": "example-2",
          "value": Object {
            "foo": "bar",
          },
        },
      ],
      "id": Any<String>,
      "mediaType": "*",
      "schema": Object {
        "$schema": "http://json-schema.org/draft-07/schema#",
        "format": "e-mail",
        "x-stoplight": Object {
          "id": Any<String>,
        },
      },
    },
  ],
  "description": "descr",
  "id": Any<String>,
  "required": true,
}
`;

exports[`params.translator translateToBodyParameter should translate to body parameter 1`] = `
Object {
  "contents": Array [
    Object {
      "examples": Array [],
      "id": Any<String>,
      "mediaType": "*",
      "schema": Object {
        "$schema": "http://json-schema.org/draft-07/schema#",
        "format": "e-mail",
        "x-stoplight": Object {
          "id": Any<String>,
        },
      },
    },
  ],
  "description": "descr",
  "id": Any<String>,
  "required": true,
}
`;

exports[`params.translator translateToHeaderParam should translate header param 1`] = `
Object {
  "description": "desc",
  "id": Any<String>,
  "name": "name",
  "required": true,
  "schema": Object {
    "$schema": "http://json-schema.org/draft-07/schema#",
    "minimum": 12,
    "type": "integer",
    "x-stoplight": Object {
      "id": Any<String>,
    },
  },
  "style": "simple",
}
`;

exports[`params.translator translateToHeaderParams should translate empty dictionary to empty array 1`] = `Array []`;

exports[`params.translator translateToHeaderParams should translate to multiple header params 1`] = `
Object {
  "description": "a description",
  "id": "117996bbb6f45",
  "name": "header-name",
  "schema": Object {
    "$schema": "http://json-schema.org/draft-07/schema#",
    "type": "string",
    "x-stoplight": Object {
      "id": Any<String>,
    },
  },
  "style": "simple",
}
`;

exports[`params.translator translateToHeaderParams should translate to multiple header params 2`] = `
Object {
  "description": "another description",
  "id": "dce8b374eb7d6",
  "name": "plain-tex",
  "schema": Object {
    "$schema": "http://json-schema.org/draft-07/schema#",
    "type": "string",
    "x-stoplight": Object {
      "id": Any<String>,
    },
  },
  "style": "simple",
}
`;

exports[`params.translator translateToHeaderParams should translate to simple header param 1`] = `
Object {
  "description": "a description",
  "id": "117996bbb6f45",
  "name": "header-name",
  "schema": Object {
    "$schema": "http://json-schema.org/draft-07/schema#",
    "type": "string",
    "x-stoplight": Object {
      "id": Any<String>,
    },
  },
  "style": "simple",
}
`;

exports[`params.translator translateToPathParameter should translate 1`] = `
Object {
  "description": "descr",
  "id": Any<String>,
  "name": "name",
  "required": true,
  "schema": Object {
    "$schema": "http://json-schema.org/draft-07/schema#",
    "type": "string",
    "x-stoplight": Object {
      "id": Any<String>,
    },
  },
  "style": "simple",
}
`;
